src/framework/preferences/js/SelfVoicingEnactor.js

```diff
@@ -92,45 +108,280 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
+     * as a node within the element's parent. The parent is also normalized to combine any adjacent textnodes.
+     *
+     * @param {String|jQuery|element} elm - element to unwrap
+     */
+    fluid.prefs.enactor.selfVoicing.unWrap = function (elm) {
+        elm = $(elm);
+
+        // remove previous marks and normalize parent to clean up textnodes
+        if (elm.length) {
+            var parent = elm.parent();
+            elm.contents().unwrap();
+            parent[0].normalize();
+        }
+    };
+
+    fluid.prefs.enactor.selfVoicing.handleUtteranceEndEvent = function (that) {
+        that.parseQueue.shift();
+        that.parseIndex = 0;
+        fluid.prefs.enactor.selfVoicing.unWrap(that.locate("mark"));
+    };
+
     // Constants representing DOM node types.
     fluid.prefs.enactor.selfVoicing.nodeType = {
         ELEMENT_NODE: 1,
         TEXT_NODE: 3
     };
 
-    // TODO: Currently only reads text nodes and alt text.
-    // This should be expanded to read other text descriptors as well.
-    fluid.prefs.enactor.selfVoicing.readFromDOM = function (that, elm) {
-        elm = $(elm);
-        var nodes = elm.contents();
-        fluid.each(nodes, function (node) {
-            if (node.nodeType === fluid.prefs.enactor.selfVoicing.nodeType.TEXT_NODE && node.nodeValue) {
-                that.tts.queueSpeech(node.nodeValue);
-            }
+    /**
+     * Tests if a string is a word, that is it has a value and is not only whitespace.
+     * inspired by https://stackoverflow.com/a/2031143
+     *
+     * @param {String} str - the String to test
+     *
+     * @return {Boolean} - `true` if a word, `false` otherwise.
+     */
+    fluid.prefs.enactor.selfVoicing.isWord = function (str) {
+        return fluid.isValue(str) && /\S/.test(str);
+    };
 
-            if (node.nodeType === fluid.prefs.enactor.selfVoicing.nodeType.ELEMENT_NODE && window.getComputedStyle(node).display !== "none") {
-                if (node.nodeName === "IMG") {
-                    var altText = node.getAttribute("alt");
-                    if (altText) {
-                        that.tts.queueSpeech(altText);
+    /**
+     * Determines if there is rendered text to in an element.
+     * Will return false in the following conditions:
+     * - elm is falsey (undefined, null, etc.)
+     * - elm's offsetHeight is 0 (e.g. display none set on itself or its parent)
+     * - elm has no text or only whitespace
+     * - elm or its parent has `aria-hidden="true"` set.
+     *
+     * @param {jQuery|element} elm - either a DOM node or a jQuery element
+     *
+     * @return {Boolean} - returns true if there is rendered text within the element and false otherwise. (See rules above)
+     */
+    fluid.prefs.enactor.selfVoicing.hasRenderedText = function (elm) {
+        elm = fluid.unwrap(elm);
+
+        return elm && !!elm.offsetHeight && fluid.prefs.enactor.selfVoicing.isWord(elm.innerText) && !$(elm).closest("[aria-hidden=\"true\"]").length;
```

Comment:
This is a very long line - I think in other contexts we have been talking about some kind of line limit like 120 or 132? I can't remember the exact figure, perhaps @simonbates can remember. The current git review interface shows me 132 characters.

Reply:
I think for GPII, we talked about using a 120 char limit.